### PR TITLE
chainring connector

### PR DESCRIPTION
**Pull request by imykolenko**
- chainring connector
 - unit tests with >80% coverage

* depends on 
    * https://github.com/Chainring-Inc/chainring-monorepo/pull/399
    * https://github.com/Chainring-Inc/chainring-monorepo/pull/409

### CODE REVIEW

File: hummingbot/connector/exchange/chainring/chainring_exchange.py [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+    def trading_rules_request_path(self):
+        return CONSTANTS.CONFIG_PATH_URL
+
+    @property
+    def trading_pairs_request_path(self):
+        return CONSTANTS.CONFIG_PATH_URL
+
+    @property
+    def check_network_request_path(self):
+        return CONSTANTS.CONFIG_PATH_URL
```
Comment (imykolenko) [inline hummingbot/connector/exchange/chainring/chainring_exchange.py:97]: all this including last trade price is implemented based on config repo

File: hummingbot/connector/exchange/chainring/chainring_exchange.py [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+            markets = exchange_info_dict.get("markets", [])
+            for market in markets:
+                try:
+                    retval.append(
+                        TradingRule(
+                            trading_pair=await self.trading_pair_associated_to_exchange_symbol(symbol=market.get("id")),
+                            min_price_increment=Decimal(market["tickSize"]),
+                            min_base_amount_increment=(Decimal(1) / Decimal("1e" + str(market["baseDecimals"]))),
+                            min_quote_amount_increment=(Decimal(1) / Decimal("1e" + str(market["quoteDecimals"])))
+                        )
```
Comment (imykolenko) [inline hummingbot/connector/exchange/chainring/chainring_exchange.py:149]: all supported settings:

```
trading_pair: str,
min_order_size: Decimal = s_decimal_0,
max_order_size: Decimal = s_decimal_max,
min_price_increment: Decimal = s_decimal_min,
min_base_amount_increment: Decimal = s_decimal_min,
min_quote_amount_increment: Decimal = s_decimal_min,
min_notional_size: Decimal = s_decimal_0,
min_order_value: Decimal = s_decimal_0,
max_price_significant_digits: Decimal = s_decimal_max,
supports_limit_orders: bool = True,
supports_market_orders: bool = True,
buy_order_collateral_token: Optional[str] = None,
sell_order_collateral_token: Optional[str] = None):
```

File: test/hummingbot/connector/exchange/chainring/test_chainring_exchange.py [hunk 3]
```diff
@@ -0,0 +1,1849 @@
...
+        return ChainringExchange(
+            client_config_map=client_config_map,
+            chainring_secret_key="0xce5715be4e423b41bb3e62bac046f9dc99041c7af3e49492e0f1b44a15de5c4b",  # noqa: mock
+            trading_pairs=[self.trading_pair],
+        )
+
+    @aioresponses()
+    def test_update_balances(self, mock_api):
+        # configure symbols response for precision transformation
+        self.configure_all_symbols_response(mock_api)
```
Comment (imykolenko) [inline test/hummingbot/connector/exchange/chainring/test_chainring_exchange.py:457]: AbstractExchangeConnectorTests.ExchangeConnectorTests provides quite a comprehensive test suite, which requires examples of all possible messages and states.

Almost all tests in this particular file overrides base tests only to add mock of all symbols response (configuration). It is needed to load symbol precisions while chainring api requires amounts to be represented in fundamental units.

File: hummingbot/connector/exchange/chainring/chainring_auth.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+        return signature['signature'].hex()
+
+    @staticmethod
+    def domain(chain_id: int, exchange_contract_address: str):
+        return {
+            'name': 'ChainRing Labs',
+            'chainId': chain_id,
+            'verifyingContract': exchange_contract_address,
+            'version': '0.0.1'
```
Comment (bholzman) [inline hummingbot/connector/exchange/chainring/chainring_auth.py:179]: Isn't the first version in semver meant to be `0.1.0`?

File: hummingbot/connector/exchange/chainring/chainring_exchange.py (hunk 2 above)
Comment (bholzman) [reply hummingbot/connector/exchange/chainring/chainring_exchange.py:149]: we do have a min order amount which can be expressed as a `min_notional_size`

File: hummingbot/connector/exchange/chainring/chainring_exchange.py [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+                path_url=f"{CONSTANTS.ORDER_PATH_URL}/{tracked_order.exchange_order_id}",
+                data=api_params,
+                is_auth_required=True,
+                limit_id=CONSTANTS.ALL_HTTP
+            )
+        except Exception as e:
+            if isinstance(e, ContentTypeError):
+                #  aiohttp seems not able to handle 204 NoContent response
+                return True
+            # if self._is_order_not_found_during_cancelation_error(e):
```
Comment (bholzman) [inline hummingbot/connector/exchange/chainring/chainring_exchange.py:321]: Does this not work?

**Review by bholzman — APPROVED** [review]
This looks great, really nice!

File: hummingbot/connector/exchange/chainring/chainring_auth.py (hunk 4 above)
Comment (imykolenko) [reply hummingbot/connector/exchange/chainring/chainring_auth.py:179]: I think you are right, it should `0.1.0`.

This value is expected by our BE and also is sent from the FE, we should change it everywhere at once.

File: hummingbot/connector/exchange/chainring/chainring_exchange.py (hunk 5 above)
Comment (imykolenko) [reply hummingbot/connector/exchange/chainring/chainring_exchange.py:322]: thanks for pointing, forgot to cleanup.
`_is_order_not_found_during_cancelation_error` is an abstract method to help bot to understand reason of failed request. During adding tests I discovered that I should not hide this error, it is handled on exchange base level and there is corresponding base level test for it.

File: hummingbot/connector/exchange/chainring/chainring_exchange.py (hunk 2 above)
Comment (imykolenko) [reply hummingbot/connector/exchange/chainring/chainring_exchange.py]: Good point, added.

**Pull request merged**